frame/grandpa/src/lib.rs

```diff
@@ -183,10 +205,81 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		/// Report some misbehavior.
-		fn report_misbehavior(origin, _report: Vec<u8>) {
-			ensure_signed(origin)?;
-			// FIXME: https://github.com/paritytech/substrate/issues/1112
+		/// Report voter equivocation/misbehavior. This method will verify the
+		/// equivocation proof and validate the given key ownership proof
+		/// against the extracted offender. If both are valid, the offence
+		/// will be reported.
+		///
+		/// FIXME: I have no clue about the weight (but we're checking two
+		/// ed25519 signatures).
+		#[weight = SimpleDispatchInfo::FixedOperational(10_000_000)]
```

Comment:
Can be ignored for now but this value is 1/10 of the block limit. I don't think this is your intention?

Reply:
I take that's way too much :relaxed: We should definitely come up with some sensitive value before merging (not necessarily too accurate for now).

Reply:
Moved validation to a `SignedExtension` so the weight is now 0.